store/tikv/txn_committer.go

```diff
@@ -338,35 +341,130 @@ func (c *txnCommitter) Commit() error {
 		}
 	}()
 
+	binlogChan := c.prewriteBinlog()
 	err := c.prewriteKeys(NewBackoffer(prewriteMaxBackoff), c.keys)
+	if binlogChan != nil {
+		binlogErr := <-binlogChan
+		if binlogErr != nil {
+			return errors.Trace(binlogErr)
+		}
+	}
 	if err != nil {
 		log.Warnf("txn commit failed on prewrite: %v, tid: %d", err, c.startTS)
+		c.rollbackBinlog()
 		return errors.Trace(err)
 	}
 
 	commitTS, err := c.store.getTimestampWithRetry(NewBackoffer(tsoMaxBackoff))
 	if err != nil {
 		log.Warnf("txn get commitTS failed: %v, tid: %d", err, c.startTS)
+		c.rollbackBinlog()
 		return errors.Trace(err)
 	}
 	c.commitTS = commitTS
 
 	if c.store.oracle.IsExpired(c.startTS, maxTxnTimeUse) {
 		err = errors.Errorf("txn takes too much time, start: %d, commit: %d", c.startTS, c.commitTS)
+		c.rollbackBinlog()
 		return errors.Annotate(err, txnRetryableMark)
 	}
 
 	err = c.commitKeys(NewBackoffer(commitMaxBackoff), c.keys)
 	if err != nil {
 		if !c.mu.committed {
 			log.Warnf("txn commit failed on commit: %v, tid: %d", err, c.startTS)
+			c.rollbackBinlog()
 			return errors.Trace(err)
 		}
 		log.Warnf("txn commit succeed with error: %v, tid: %d", err, c.startTS)
 	}
+	c.writeCommitBinlog()
```

Comment:
Why not put `c.writeCommitBinlog()` and `c.rollbackBinlog()` in `tikvTxn.Commit()`?